src/architect/project/project.utils.ts

```diff
@@ -0,0 +1,319 @@
+import fs from 'fs-extra';
+import yaml from 'js-yaml';
+import inquirer from 'inquirer';
+import fetch from 'node-fetch';
+import { Dictionary } from '../../dependency-manager/utils/dictionary';
+import { ComponentSpec } from '../../';
+import { ServiceSpec } from '../../dependency-manager/spec/service-spec';
+
+const download = require('download-git-repo');
+
+interface GitHubRepo {
+  name: string,
+  full_name: string,
+}
+
+interface Selection {
+  name: string,
+  architect_yml: Dictionary<any>,
+  depends_on: string,
+}
+
+const GITHUB_BRANCH = 'main';
+
+class Queue {
+  items: any[];
+
+  constructor(...params: any[]) {
+    this.items = [...params];
+  }
+
+  enqueue(item: any) {
+    this.items.push(item);
+  }
+
+  dequeue() {
+    return this.items.shift();
+  }
+
+  getItems() {
+    return this.items;
+  }
+
+  hasItem(key: string) {
+    return this.items.some(item => item.hasOwnProperty(key));
+  }
+
+  size() {
+    return this.items.length;
+  }
+}
+
+export default class ProjectUtils {
+
+  static async getGitHubRepos(url: string): Promise<GitHubRepo[]>{
+    const res_json = await this.fetchJsonFromGitHub('https://api.github.com/orgs/architect-templates/repos');
+    const repos = res_json.map((item: any) => {
+      return {
+        name: item.name,
+        full_name: item.full_name,
+      };
+    });
+    return repos;
+  }
+
+  static async fetchJsonFromGitHub(url: string): Promise<any> {
+    const response = await fetch(url);
+    if (!response.ok) {
+      throw new Error(`Failed to fetch ${url}`);
+    }
+    return await response.json();
+  }
+
+  static async fetchYamlFromGitHub(url: string): Promise<ComponentSpec> {
+    const component_spec = await fetch(url)
+      .then(res => res.blob())
+      .then(blob => blob.text())
+      .then(yaml_as_string => yaml.load(yaml_as_string) as ComponentSpec)
+      .catch(err => {
+        throw new Error(`Failed to fetch ${url}`);
+      });
+    return component_spec;
+  }
+
+  static async getSelections(chosen_project: string): Promise<Dictionary<Selection>> {
+    const queue = new Queue();
+    const type_map: Dictionary<any> = {};
+    const dependency_map: Dictionary<any> = {};
+    const component_map: Dictionary<any> = {};
+
+    // find project repository in our GitHub
+    const github_repos = await this.getGitHubRepos('https://api.github.com/orgs/architect-templates/repos');
+    const start_project = github_repos.find(repo => repo.name === chosen_project.toLowerCase());
+    if (!start_project) {
+      const repo_names = github_repos.map(repo => repo.name).join(', ');
+      throw new Error(`Cannot find project '${chosen_project}'. Available projects are: ${repo_names}.`);
+    }
+
+    // store component spec of the chosen project
+    const yaml_url = 'https://raw.githubusercontent.com/' + start_project.full_name + `/${GITHUB_BRANCH}/architect.yml`;
```

Comment:
So we kind of duplicate this logic. The first time we initalize the queue and the second time to keep adding new choices to the queue. If the queue instead was just a list of repos to add we could move all the logic into the while loop and not duplicate it.

So we just take the starting component name add it to the queue and then enter the loop. All logic from there just happens in the loop.